From: Gary Lin <glin@suse.com>
Date: Wed, 30 Nov 2016 15:11:27 +0800
Subject: [PATCH] libata: remove n_tags to avoid kABI breakage
Patch-mainline: Never, SLE11-SP4 specific
References: bsc#871728

Introduing n_tags to struct ata_host causes kABI breakage in SLE11-SP4.
Instead of adding n_tags, a conditional check of ata_port->scsi_host is
added to avoid the potential crash.

If ata_port->scsi is uninitialized, it's a SAS controller, and the default
queue depth, ATA_MAX_QUEUE - 1, is assigned to max_queue. Otherwise, it's
a !SAS controller, and we clamp ata_port->scsi_host->can_queue to make sure
it won't exceed ATA_MAX_QUEUE.

Signed-off-by: Gary Lin <glin@suse.com>
---
 drivers/ata/libata-core.c | 15 +++++++++++----
 include/linux/libata.h    |  1 -
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/ata/libata-core.c b/drivers/ata/libata-core.c
index 9b2e50e..cbb38ac 100644
--- a/drivers/ata/libata-core.c
+++ b/drivers/ata/libata-core.c
@@ -4746,9 +4746,19 @@ void swap_buf_le16(u16 *buf, unsigned int buf_words)
 static struct ata_queued_cmd *ata_qc_new(struct ata_port *ap)
 {
 	struct ata_queued_cmd *qc = NULL;
-	unsigned int max_queue = ap->host->n_tags;
+	unsigned int max_queue;
 	unsigned int i, tag;
 
+	/*
+	 * Since SAS controllers doesn't initialize ->scsi_host, we have to
+	 * check if ata_port->scsi_host exists or not.
+	 */
+	if (ap->scsi_host)
+		max_queue = clamp(ap->scsi_host->can_queue, 1,
+				  ATA_MAX_QUEUE - 1);
+	else
+		max_queue = ATA_MAX_QUEUE - 1;
+
 	/* no command while frozen */
 	if (unlikely(ap->pflags & ATA_PFLAG_FROZEN))
 		return NULL;
@@ -6035,7 +6045,6 @@ void ata_host_init(struct ata_host *host, struct device *dev,
 {
 	spin_lock_init(&host->lock);
 	mutex_init(&host->eh_mutex);
-	host->n_tags = ATA_MAX_QUEUE - 1;
 	host->dev = dev;
 	host->ops = ops;
 }
@@ -6117,8 +6126,6 @@ int ata_host_register(struct ata_host *host, struct scsi_host_template *sht)
 {
 	int i, rc;
 
-	host->n_tags = clamp(sht->can_queue, 1, ATA_MAX_QUEUE - 1);
-
 	/* host must have been started */
 	if (!(host->flags & ATA_HOST_STARTED)) {
 		dev_printk(KERN_ERR, host->dev,
diff --git a/include/linux/libata.h b/include/linux/libata.h
index b5388f9..50f74e5 100644
--- a/include/linux/libata.h
+++ b/include/linux/libata.h
@@ -533,7 +533,6 @@ struct ata_host {
 	struct device 		*dev;
 	void __iomem * const	*iomap;
 	unsigned int		n_ports;
-	unsigned int		n_tags;			/* nr of NCQ tags */
 	void			*private_data;
 	struct ata_port_operations *ops;
 	unsigned long		flags;
-- 
2.10.2

